File: aula09/Pair.lua
require 'oop'

--Pair = newClass()
class 'Pair'


function Pair.init(self, x, y)
    self.x = x
    self.y = y
end

function Pair:__tostring()
    return string.format('Pair(%s, %s)', self.x, self.y)
end


function Pair:abs()
    return math.sqrt(self.x^2 + self.y^2)
end

function Pair:scale(factor)
    return Pair(self.x * factor, self.y * factor)
end


--public void translate(double x, double y) {
--    this.x += x;
--    this.y += y; // this.y = this.y + y;
--}

File: ex_pair/java/Pair.java
public class Pair {
    private double x;
    private double y;

    public Pair(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double abs() {
        return Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2));
    }

    public String toString() {
        return "(" + x + ", " + y + ")";
    }

    // region getters

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    // endregion


    // region setters

    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    // endregion

    // region Mudando estado do objeto

    public void scale(double factor) {
        x *= factor;
        y *= factor;
    }

    public void translate(double x, double y) {
        this.x += x;
        this.y += y; // this.y = this.y + y;
    }

    public void translate(Pair p) {
        this.x += p.x;
        this.y += p.y;
    }

    // endregion

}